contracts/TwabRewards.sol

```diff
@@ -0,0 +1,419 @@
+            "TwabRewards/epoch-not-over"
+        );
+
+        ITicket _ticket = ITicket(_promotion.ticket);
+
+        uint256 _averageBalance = _ticket.getAverageBalanceBetween(
+            _user,
+            uint64(_epochStartTimestamp),
+            uint64(_epochEndTimestamp)
+        );
+
+        uint64[] memory _epochStartTimestamps = new uint64[](1);
+        _epochStartTimestamps[0] = uint64(_epochStartTimestamp);
+
+        uint64[] memory _epochEndTimestamps = new uint64[](1);
+        _epochEndTimestamps[0] = uint64(_epochEndTimestamp);
+
+        uint256[] memory _averageTotalSupplies = _ticket.getAverageTotalSuppliesBetween(
+            _epochStartTimestamps,
+            _epochEndTimestamps
+        );
+
+        return (_promotion.tokensPerEpoch * _averageBalance) / _averageTotalSupplies[0];
+    }
+
+    /**
+        @notice Get the total amount of tokens left to be rewarded.
+        @param _promotionId Promotion id to get the total amount of tokens left to be rewarded for
+        @return Amount of tokens left to be rewarded
+     */
+    function _getRemainingRewards(uint256 _promotionId) internal view returns (uint256) {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint256 _numberOfEpochs = _promotion.numberOfEpochs;
+        uint256 _tokensPerEpoch = _promotion.tokensPerEpoch;
+        uint256 _currentEpochId = _getCurrentEpochId(_promotionId);
+
+        if (_currentEpochId == 0) {
+            return _tokensPerEpoch * _numberOfEpochs;
+        } else if (_numberOfEpochs <= _currentEpochId + 1) {
+            return 0;
+        } else {
+            // _tokensPerEpoch * _numberOfEpochsLeft
+            return _tokensPerEpoch * (_numberOfEpochs - (_currentEpochId + 1));
+        }
+    }
+
+    /**
+        @notice Set boolean value for a specific epoch.
+        @dev Bits are stored in a uint256 from right to left.
+        Let's take the example of the following 8 bits word. 0110 0111
+        To set the boolean value to 0 for the epoch id 2, we need to create a mask by shifting 1 to the left by 2 bits.
+        We get: 0000 0001 << 2 = 0000 0100
+        We then OR the mask with the word to set the value.
+        We get: 0110 0111 | 0000 0100 = 0110 0111
+        To set the boolean value to 0 for the epoch id 2, we need to create a mask by shifting 1 to the left by 2 bits and then inverting it.
+        We get: 0000 0001 << 2 = ~(0000 0100) = 1111 1011
+        We then AND the mask with the word to clear the value.
+        We get: 0110 0111 & 1111 1011 = 0110 0011
+        @param _epochs Tightly packed epoch ids with their boolean values
+        @param _epochId Id of the epoch to set the boolean for
+        @param _value Boolean value to set
+        @return Tightly packed epoch ids with the newly boolean value set
+    */
+    function _setClaimedEpoch(
+        uint256 _epochs,
+        uint256 _epochId,
+        bool _value
```

Comment:
why are we passing a boolean here for _value when it will always be true?

Reply:
what you will want to do here is:
```
function _setClaimedEpoch(uint256 _epochs, uint256 _epochId, address _user){
    _claimedEpochs[_epochId][_user] = _epochs | (uint256(1) << _epochId)
}
```

Reply:
but actually if we are supporting multiple claims at the same time in a loop this will be inefficient since we are SSTORE'ing here.

Reply:
What we really want is an `_updateClaimedEpoch()` function that we can call within the call site loop and keep the return result of this function in memory. Then outside the calling loop we write it to storage, only doing one SSTORE per multiple claims

Reply:
https://github.com/pooltogether/v4-periphery/pull/10#issue-1069920459

Reply:
Merged: https://github.com/pooltogether/v4-periphery/pull/8/commits/6793babebb941339bfe67b55b56e3e3bc70add25